R/metab_simple.R

```diff
@@ -48,31 +54,39 @@ metab_simple <- function(data, ...) {
   #' 
   #' @param params a vector of length 2, where the first element is GPP and the second element is ER (both mg/L/d)
   #' @param 
-#   onestation_likelihood <- function(params, DO, DO.deficit, oxy, light, z, bp, ts, K) {
-#     
-#     GPP <- params[1]
-#     ER <- params[2]
-#     
-#     DO.modeled <- numeric(nrow(data))
-#     DO.modeled[1] <- oxy[1]
-#     for (i in 2:length(DO)) {
-#       DO.modeled[i] <- 
-#         DO.modeled[i-1] +
-#         (GPP/z)*(light[i]/sum(light)) + 
-#         ER * ts / z + 
-#         Kcor(temp[i],K)*ts*DO.deficit 
-#     }
-#   
-#     ##below is MLE calculation
-#     sqdiff<-(oxy-metab)^2 
-#     length(oxy)*(log(((sum(sqdiff)/length(oxy))^0.5)) +0.5*log(6.28))   + ((2*sum(sqdiff)/length(oxy))^-1)*sum(sqdiff)
-#   }
-  
+  onestation_negloglik <- function(params, DO.obs, DO.deficit, depth, k.O2, frac.GPP, frac.ER, frac.D) {
+    
+    # Parse params vector (passed from nlm)
+    GPP <- params[1]
+    ER <- params[2]
+    
+    # Model DO with given params
+    DO.mod <- numeric(length(DO.obs))
+    DO.mod[1] <- DO.obs[1]
+    DO.delta <- 
+      GPP * frac.GPP / depth + 
+      ER * frac.ER / depth + 
+      k.O2 * DO.deficit * frac.D # Bob - pros & cons of using DO.mod[t-1] instead of DO.obs[t] or DO.obs[t-1] (i.e., from DO.deficit) here?
```

Comment:
@jread-usgs pointed this discussion out to me – so I'll chime in.

I've thought about this same issue myself. If I'm following correctly, I agree with Bob.

So your parameter guesses are pretty bad, and your modeled DO at the last time step was something like 1000% of saturation, but the observed DO was 90%. You'd want the model to say there was a huge eflux of oxygen from the water to the atmosphere, not a minor influx, right? That's how I've thought of it.

when I do the "ols" method, I actually disobey this rule. In those models, there's something to the effect of `d[t] = d[t-1] + XB + F[t-1] +error[t]`. Using `lm()` in R, you can't simply do this calculation dynamically (even though these are dynamical models). So for my response variable I'll have `delta <- diff(d)-head(F, -1)`, and I'll also remove the last observation from all the X (`head(X,-1)`). Then my regression is something like `lm(delta ~ X)`.

In my experience, the consequences of these different approaches are relatively minor when it comes to your final metabolism estimate. Consider, e.g., estimating `b` in the following: `x[t] = b*x[t-1] + error[t]`. This is an AR(1) model. It is not uncommon to estimate first-order autocorrelation by regressing a variable against itself: `x.lag <- embed(x,2); lm(x.lag[,1] ~ x.lag[,2])`. In fact, in R do `?ar` – you'll see a variety of ways of estimating the ar coefficient, and these vary greatly in complexity. One of the methods is `ar.ols`.

Raising points like this is why I often appreciate state space representations of these dynamic models. For example, if in R you do `?arima` you'll see a Kalman filter is used to fit the arima models (even when the arima model just boils down to an AR1). For LakeMetabolizer, we use [metab.kalman](https://github.com/GLEON/LakeMetabolizer/blob/master/R/metab.kalman.R) for a similar approach, although we also have a Bayesian model for the state space representation.